pkg/controller/keycloakclient/keycloakclient_reconciler.go

```diff
@@ -43,9 +43,85 @@ func (i *KeycloakClientReconciler) Reconcile(state *common.ClientState, cr *kc.K
 		desired.AddAction(i.getUpdatedClientSecretState(state, cr))
 	}
 
+	i.ReconcileRoles(state, cr, &desired)
+
 	return desired
 }
 
+func (i *KeycloakClientReconciler) ReconcileRoles(state *common.ClientState, cr *kc.KeycloakClient, desired *common.DesiredClusterState) {
+	// delete existing roles for which no desired role is found that (matches by ID OR has no ID but matches by name)
+	// this implies that specifying a role with matching name but different ID will result in deletion (and re-creation)
+	rolesDeleted, _ := roleDifferenceIntersection(state.Roles, cr.Spec.Roles)
+	for _, role := range rolesDeleted {
+		desired.AddAction(i.getDeletedClientRoleState(state, cr, role.DeepCopy()))
+	}
+
+	// update with desired roles that can be matched to existing roles and have an ID set, this includes all renames
+	// note down all renames
+	existingRoleByID := make(map[string]kc.RoleRepresentation)
+	for _, role := range state.Roles {
+		existingRoleByID[role.ID] = role
+	}
+	renamedRolesOldNames := make(map[string]bool)
+	_, rolesMatching := roleDifferenceIntersection(cr.Spec.Roles, state.Roles)
+	for _, role := range rolesMatching {
+		if role.ID != "" {
+			oldRole := existingRoleByID[role.ID]
+			desired.AddAction(i.getUpdatedClientRoleState(state, cr, role.DeepCopy(), oldRole.DeepCopy()))
+			if role.Name != oldRole.Name {
+				renamedRolesOldNames[oldRole.Name] = true
+			}
+		}
+	}
+
+	// seemingly matching roles without an ID can either be regular updates
+	// or re-creations after renames (not deletions)
+	// note that duplicate role names are impossible thanks to +listType=map
+	for _, role := range rolesMatching {
+		if role.ID == "" {
+			if _, contains := renamedRolesOldNames[role.Name]; contains {
+				desired.AddAction(i.getCreatedClientRoleState(state, cr, role.DeepCopy()))
+			} else {
+				desired.AddAction(i.getUpdatedClientRoleState(state, cr, role.DeepCopy(), role.DeepCopy()))
+			}
+		}
+	}
```

Comment:
I'm not sure I understand why would be ID missing from the CR when doing an update. Or why we would be re-creating the role after rename. The user shouldn't manually remove the ID from the CR. Or maybe I'm just lacking imagination on Monday morning... :D

Reply:
Not something I think most people would need, in fact I didn't even consider renaming initially.
But going from name matching to [ID then name] matching comes with a couple of corner cases and instead of failing with weird errors in the operator log I implemented what would be needed to get to the desired state.
Note that the user doesn't necessarily need to "remove" the ID from the CR, if they don't specify it initially we will still run into this because the read state contains IDs (because KC would just generate them) whereas the CR does not (and maybe never did).

Reply:
Ok, thanks for clarifying. Good point with the CR missing the ID.